smithy-jmespath/src/main/java/software/amazon/smithy/jmespath/evaluation/Evaluator.java

```diff
@@ -0,0 +1,346 @@
+            flattened.add(val);
+        }
+        return flattened.build();
+    }
+
+    @Override
+    public T visitFunction(FunctionExpression functionExpression) {
+        Function function = FunctionRegistry.lookup(functionExpression.getName());
+        if (function == null) {
+            throw new JmespathException(JmespathExceptionType.UNKNOWN_FUNCTION, functionExpression.getName());
+        }
+        List<FunctionArgument<T>> arguments = new ArrayList<>();
+        for (JmespathExpression expr : functionExpression.getArguments()) {
+            if (expr instanceof ExpressionTypeExpression) {
+                arguments.add(FunctionArgument.of(runtime, ((ExpressionTypeExpression) expr).getExpression()));
+            } else {
+                arguments.add(FunctionArgument.of(runtime, visit(expr)));
+            }
+        }
+        return function.apply(runtime, arguments);
+    }
+
+    @Override
+    public T visitField(FieldExpression fieldExpression) {
+        return runtime.value(current, runtime.createString(fieldExpression.getName()));
+    }
+
+    @Override
+    public T visitIndex(IndexExpression indexExpression) {
+        int index = indexExpression.getIndex();
+        if (!runtime.is(current, RuntimeType.ARRAY)) {
+            return runtime.createNull();
+        }
+        // TODO: Capping at int here unnecessarily
+        // Perhaps define intLength() and return -1 if it doesn't fit?
+        // Although technically IndexExpression should be using a Number instead of an int in the first place
+        int length = runtime.length(current).intValue();
+        // Negative indices indicate reverse indexing in JMESPath
+        if (index < 0) {
+            index = length + index;
+        }
+        if (length <= index || index < 0) {
+            return runtime.createNull();
+        }
+        return runtime.element(current, runtime.createNumber(index));
+    }
+
+    @Override
+    public T visitLiteral(LiteralExpression literalExpression) {
+        if (literalExpression.isNumberValue()) {
+            return runtime.createNumber(literalExpression.expectNumberValue());
+        } else if (literalExpression.isArrayValue()) {
+            JmespathRuntime.ArrayBuilder<T> result = runtime.arrayBuilder();
+            for (Object item : literalExpression.expectArrayValue()) {
+                result.add(visit(LiteralExpression.from(item)));
+            }
+            return result.build();
+        } else if (literalExpression.isObjectValue()) {
+            JmespathRuntime.ObjectBuilder<T> result = runtime.objectBuilder();
+            for (Map.Entry<String, Object> entry : literalExpression.expectObjectValue().entrySet()) {
+                T key = runtime.createString(entry.getKey());
+                T value = visit(LiteralExpression.from(entry.getValue()));
+                result.put(key, value);
+            }
+            return result.build();
+        } else if (literalExpression.isStringValue()) {
```

Comment:
nit but string and boolean are likely more common than array and object literals